internal/process/process.go

```diff
@@ -0,0 +1,162 @@
+package process
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"os/exec"
+	"sync"
+	"time"
+)
+
+// Process is a long-running process manager that will restart processes if they stop.
+type Process struct {
+	Cmd          *exec.Cmd
+	Stdin        io.WriteCloser
+	Stdout       io.ReadCloser
+	Stderr       io.ReadCloser
+	ReadStdoutFn func(io.Reader)
+	ReadStderrFn func(io.Reader)
+	RestartDelay time.Duration
+
+	cancel     context.CancelFunc
+	mainLoopWg sync.WaitGroup
+}
+
+// New creates a new process wrapper
+func New(command []string) (*Process, error) {
+	p := &Process{
+		RestartDelay: 5 * time.Second,
+	}
+	if len(command) > 1 {
+		p.Cmd = exec.Command(command[0], command[1:]...)
+	} else {
+		p.Cmd = exec.Command(command[0])
+	}
+	var err error
+	p.Stdin, err = p.Cmd.StdinPipe()
+	if err != nil {
+		return nil, fmt.Errorf("Error opening stdin pipe: %s", err)
+	}
+
+	p.Stdout, err = p.Cmd.StdoutPipe()
+	if err != nil {
+		return nil, fmt.Errorf("Error opening stdout pipe: %s", err)
+	}
+
+	p.Stderr, err = p.Cmd.StderrPipe()
+	if err != nil {
+		return nil, fmt.Errorf("Error opening stderr pipe: %s", err)
+	}
+
+	return p, nil
+}
+
+// Start the process
+func (p *Process) Start() error {
+	p.mainLoopWg.Add(1)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	p.cancel = cancel
+
+	if err := p.cmdStart(); err != nil {
+		return err
+	}
+
+	go func() {
+		if err := p.cmdLoop(ctx); err != nil {
+			log.Printf("Process quit with message: %s", err.Error())
+		}
+		p.mainLoopWg.Done()
+	}()
+
+	return nil
+}
+
+func (p *Process) Stop() {
+	if p.cancel != nil {
+		p.cancel()
+	}
+	p.mainLoopWg.Wait()
+}
+
+func (p *Process) cmdStart() error {
+	log.Printf("Starting process: %s %s", p.Cmd.Path, p.Cmd.Args)
+
+	if err := p.Cmd.Start(); err != nil {
+		return fmt.Errorf("Error starting process: %s", err)
+	}
+
+	return nil
+}
+
+// cmdLoop watches an already running process, restarting it when appropriate.
+func (p *Process) cmdLoop(ctx context.Context) error {
```

Comment:
Setup this function to return only once the process has exited.  This way you know when the process is finished and there won't be any calls to ReadStdoutFn later on.

Reply:
This function looks more complicated than it is. rewriting it.